src/libopensc/card-esteid2018.c

```diff
@@ -0,0 +1,327 @@
+
+	LOG_FUNC_CALLED(card->ctx);
+	if (data == NULL || out == NULL || datalen > SIGNATURE_PAYLOAD_SIZE)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+
+	env = &priv->sec_env;
+	// left-pad if necessary
+	memcpy(&sbuf[SIGNATURE_PAYLOAD_SIZE - datalen], data, MIN(datalen, SIGNATURE_PAYLOAD_SIZE));
+	memset(sbuf, 0x00, SIGNATURE_PAYLOAD_SIZE - datalen);
+	datalen = SIGNATURE_PAYLOAD_SIZE;
+
+	switch (env->key_ref[0]) {
+	case 1: /* authentication key */
+		sc_format_apdu_ex(card, &apdu, 0x88, 0, 0, sbuf, datalen, out, MIN(256, outlen));
+		break;
+	default:
+		sc_format_apdu_ex(card, &apdu, 0x2A, 0x9E, 0x9A, sbuf, datalen, out, MIN(256, outlen));
+	}
+
+	SC_TRANSMIT_TEST_RET(card, apdu, "PSO CDS/INTERNAL AUTHENTICATE failed");
+
+	LOG_FUNC_RETURN(card->ctx, apdu.resplen);
+}
+
+static int esteid_get_pin_remaining_tries(sc_card_t *card, int pin_reference) {
+	unsigned char get_pin_info[] = {0x4D, 0x08, 0x70, 0x06, 0xBF, 0x81, 0xFF, 0x02, 0xA0, 0x80};
+
+	struct sc_apdu apdu;
+	unsigned char apdu_resp[SC_MAX_APDU_BUFFER_SIZE];
+	LOG_FUNC_CALLED(card->ctx);
+
+	// We don't get the file information here, so we need to be ugly
+	if (pin_reference == 1 || pin_reference == 2) {
+		LOG_TEST_RET(card->ctx, esteid_select(card, 0x00, 0x3F, 0x00), "Cannot select MF");
+	} else if (pin_reference == 0x85) {
+		LOG_TEST_RET(card->ctx, esteid_select_file(card, &adf2, NULL), "Cannot select QSCD AID");
+	} else {
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	}
+
+	get_pin_info[6] = pin_reference & 0x0F; // mask out local/global
+	// XXX: handling of default Le of 0x00 could be easier
+	sc_format_apdu_ex(card, &apdu, 0xCB, 0x3F, 0xFF, get_pin_info, sizeof(get_pin_info), apdu_resp, MIN(256, sizeof(apdu_resp)));
+	SC_TRANSMIT_TEST_RET(card, apdu, "GET DATA(pin info) failed");
+
+	// XXX: sc_asn1_find_tag with the following payload (to get to tag 0x9B):
+	// https://lapo.it/asn1js/#cB6_gQEaoBiaAQObAQOhEIwG8wAAc0MAnAbzAABzQwA
+	return apdu_resp[13];
+}
+
+static int esteid_pin_cmd(sc_card_t *card, struct sc_pin_cmd_data *data, int *tries_left) {
+	int r;
+	struct sc_pin_cmd_data tmp;
+	LOG_FUNC_CALLED(card->ctx);
+	sc_log(card->ctx, "PIN CMD is %d", data->cmd);
+	if (data->cmd == SC_PIN_CMD_GET_INFO) {
+		sc_log(card->ctx, "SC_PIN_CMD_GET_INFO for %d", data->pin_reference);
+		r = esteid_get_pin_remaining_tries(card, data->pin_reference);
+		LOG_TEST_RET(card->ctx, r, "GET DATA(pin info) failed");
+
+		data->pin1.tries_left = r;
+		data->pin1.max_tries = -1; // "no support, which means the one set in PKCS#15 emulation sticks
+		data->pin1.logged_in = SC_PIN_STATE_UNKNOWN;
+		LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+	} else if (data->cmd == SC_PIN_CMD_UNBLOCK) {
+		// Verify PUK, then issue UNBLOCK
+		// VERIFY
+		memcpy(&tmp, data, sizeof(struct sc_pin_cmd_data));
+		tmp.cmd = SC_PIN_CMD_VERIFY;
+		tmp.pin_reference = 0x02; // hardcoded, ugly
+		tmp.pin2.len = 0;
+		r = iso_ops->pin_cmd(card, &tmp, tries_left);
+		sc_mem_clear(&tmp, sizeof(tmp));
```

Comment:
no need to clear the memory, because the actual PINs (`tmp.pin1.data`/`tmp.pin2.data`) are only pointers.